File: 5_HandsOn_Tutorials/5.8_Message_Queues/5.8.1_RabbitMQ/02_exchanges_routing/topic_consumer.py
#!/usr/bin/env python3
"""
RabbitMQ Topic Exchange Consumer

Subscribes using routing key patterns with wildcards:
  * (star) - matches exactly one word
  # (hash) - matches zero or more words

Examples:
  *.error.*        - All error messages
  app.#            - All app facility messages
  #.database       - All database component messages
  kernel.*.memory  - All kernel memory messages
"""

import pika
import json
import sys


def create_connection():
    """Create connection to RabbitMQ"""
    credentials = pika.PlainCredentials('admin', 'admin123')
    parameters = pika.ConnectionParameters(
        host='localhost',
        port=5672,
        credentials=credentials
    )
    return pika.BlockingConnection(parameters)


def callback(ch, method, properties, body):
    """Process received messages"""
    try:
        message = json.loads(body.decode('utf-8'))
        routing_key = method.routing_key

        print(f"\n{'='*60}")
        print(f"📨 Received message:")
        print(f"   Routing Key: {routing_key}")
        print(f"   Text: {message['text']}")
        print(f"   Facility: {message['facility']}")
        print(f"   Severity: {message['severity']}")
        print(f"   Component: {message['component']}")
        print(f"   Timestamp: {message['timestamp']}")
        print(f"{'='*60}")

        ch.basic_ack(delivery_tag=method.delivery_tag)
    except Exception as e:
        print(f"✗ Error processing message: {e}")
        ch.basic_nack(delivery_tag=method.delivery_tag, requeue=False)


def main():
    """Main function"""
    exchange_name = 'topic_logs'

    # Get binding patterns from command line or use defaults
    if len(sys.argv) > 1:
        binding_keys = sys.argv[1:]
    else:
        # Default: subscribe to all error and critical messages
        binding_keys = ['*.error.*', '*.critical.*']

    # Connect to RabbitMQ
    connection = create_connection()
    channel = connection.channel()

    # Declare exchange
    channel.exchange_declare(
        exchange=exchange_name,
        exchange_type='topic',
        durable=True
    )

    # Create exclusive queue
    result = channel.queue_declare(queue='', exclusive=True)
    queue_name = result.method.queue

    # Bind queue for each pattern
    print("📌 Binding patterns:")
    for binding_key in binding_keys:
        channel.queue_bind(
            exchange=exchange_name,
            queue=queue_name,
            routing_key=binding_key
        )
        print(f"   ✓ {binding_key}")

    # Set up consumer
    channel.basic_qos(prefetch_count=1)
    channel.basic_consume(
        queue=queue_name,
        on_message_callback=callback,
        auto_ack=False
    )

    print(f"\n📡 Waiting for messages matching patterns...")
    print("   Press CTRL+C to exit\n")

    try:
        channel.start_consuming()
    except KeyboardInterrupt:
        print('\n✗ Interrupted')
        channel.stop_consuming()
    finally:
        connection.close()


if __name__ == '__main__':
    try:
        main()
    except Exception as e:
        print(f'✗ Error: {e}')
        sys.exit(1)

File: 5_HandsOn_Tutorials/5.8_Message_Queues/5.8.1_RabbitMQ/02_exchanges/topic_consumer.ts
#!/usr/bin/env ts-node
import * as amqp from 'amqplib';
import { Connection, Channel, ConsumeMessage } from 'amqplib';

/**
 * Topic message interface
 */
interface TopicMessage {
    routing_key: string;
    message: string;
    timestamp: string;
}

/**
 * Consume messages from topic exchange with pattern matching
 */
async function consumeTopicMessages(bindingKeys: string[]): Promise<void> {
    let connection: Connection | null = null;
    let channel: Channel | null = null;

    try {
        const connectionUrl = 'amqp://admin:password@localhost:5672/';
        connection = await amqp.connect(connectionUrl);
        channel = await connection.createChannel();

        // Declare exchange
        const exchangeName = 'logs_topic';
        await channel.assertExchange(exchangeName, 'topic', {
            durable: true
        });

        // Create exclusive queue
        const queueResult = await channel.assertQueue('', { exclusive: true });
        const queueName = queueResult.queue;

        // Bind queue with each pattern
        for (const bindingKey of bindingKeys) {
            await channel.bindQueue(queueName, exchangeName, bindingKey);
            console.log(`Bound to pattern: ${bindingKey}`);
        }

        await channel.prefetch(1);

        console.log(`\n⏳ Waiting for messages matching ${bindingKeys.join(', ')}. Press CTRL+C to exit\n`);

        // Start consuming
        await channel.consume(
            queueName,
            (msg: ConsumeMessage | null) => {
                if (msg !== null) {
                    try {
                        const message: TopicMessage = JSON.parse(msg.content.toString());
                        console.log(`✓ [${msg.fields.routingKey}] ${message.message}`);
                        console.log(`  Timestamp: ${message.timestamp}`);
                        channel!.ack(msg);
                    } catch (error) {
                        console.error('Error processing message:', error);
                        channel!.nack(msg, false, false);
                    }
                }
            },
            {
                noAck: false
            }
        );

        // Handle graceful shutdown
        process.on('SIGINT', async () => {
            console.log('\n✓ Consumer stopped');
            try {
                if (channel) await channel.close();
                if (connection) await connection.close();
                process.exit(0);
            } catch (error) {
                console.error('Error during shutdown:', error);
                process.exit(1);
            }
        });

    } catch (error) {
        console.error('Error setting up consumer:', error);

        try {
            if (channel) await channel.close();
            if (connection) await connection.close();
        } catch (closeError) {
            console.error('Error closing connection:', closeError);
        }

        throw error;
    }
}

// Main execution
if (require.main === module) {
    const args = process.argv.slice(2);

    if (args.length < 1) {
        console.log('Usage: ts-node topic_consumer.ts <pattern1> [pattern2] ...');
        console.log('Patterns:');
        console.log('  * (star) - matches exactly one word');
        console.log('  # (hash) - matches zero or more words');
        console.log('\nExamples:');
        console.log('  ts-node topic_consumer.ts "*.critical"       # All critical messages');
        console.log('  ts-node topic_consumer.ts "kern.*"          # All kernel messages');
        console.log('  ts-node topic_consumer.ts "app.#"           # All app messages');
        console.log('  ts-node topic_consumer.ts "#"               # All messages');
        process.exit(1);
    }

    const bindingKeys = args;
    consumeTopicMessages(bindingKeys).catch(error => {
        console.error('Fatal error:', error);
        process.exit(1);
    });
}

export { consumeTopicMessages, TopicMessage };
